tls/s2n_server_hello_retry.c

```diff
@@ -93,14 +115,22 @@ int s2n_server_hello_retry_recv(struct s2n_connection *conn)
         new_key_share_requested = (kem_group != conn->kex_params.client_kem_group_params.kem_group);
     }
 
-    /*
+    /**
      *= https://tools.ietf.org/rfc/rfc8446#section-4.1.4
      *# Clients MUST abort the handshake with an
      *# "illegal_parameter" alert if the HelloRetryRequest would not result
      *# in any change in the ClientHello.
-     */
+     *
+     *= https://tools.ietf.org/rfc/rfc8446#4.2.8
+     *= type=exception
+     *= reason=Permit HRRs to leave the selected_group field unmodified if sent due to rejecting early data
+     *# If either of these checks fails, then
+     *# the client MUST abort the handshake with an "illegal_parameter"
+     *# alert.
```

Comment:
Even with early data, don't you need to offer a correct selected group? Why is there an exception for early data?

Reply:
Yes, a correct selected group still needs to be offered. This exception is for this line:

https://github.com/aws/s2n-tls/blob/5725d38f5d8971dd941601b94adb713a7718b343/tls/s2n_server_hello_retry.c#L107-L108

Without the exception, I think there should not be a check for `conn->early_data_state == S2N_EARLY_DATA_REJECTED`. It seems like this condition is here to allow for HRRs to not modify the selected_group field if the HRR was sent only for the purpose of rejecting early data. I think this makes sense, because if you're only rejecting early data it wouldn't make sense to also need to update the selected_group. However, this condition does not seem to be present in the RFC. 

Here is the relevant part:
```
   Upon receipt of this extension in a HelloRetryRequest, the client
   MUST verify that (1) the selected_group field corresponds to a group
   which was provided in the "supported_groups" extension in the
   original ClientHello and (2) the selected_group field does not
   correspond to a group which was provided in the "key_share" extension
   in the original ClientHello.  If either of these checks fails, then
   the client MUST abort the handshake with an "illegal_parameter"
   alert.
```

Checks (1) and (2) must both succeed. And, according to check (2), `new_key_share_requested` should always be true. So, the exception I added is for this statement: 

```
   If either of these checks fails, then
   the client MUST abort the handshake with an "illegal_parameter"
   alert.
```

Since we allow check (2) to be false if early data was rejected.

Let me know if I'm misunderstanding something or if we actually should not be making this exception!

Reply:
I think the misunderstanding comes from the fact that the HelloRetryRequest doesn't _have_ to include a keyshare extension. A server who is using a HRR to reject early data would not include a keyshare extension in that request. So the client doesn't need to make an exception in this case. Basically, if the server includes a keyshare extension in its HRR, it _has_ to be asking for a keyshare change, otherwise the server can just leave that extension out.

So what I'm saying is, at minimum, there is no exception to the RFC. However, the existing code might need to be rewritten for the case where the server uses HRR to reject early data. 😪 @lrstewart do you know if we have a test for this?

This paragraph shows that the keyshare extension is optional in a HRR:
> 
   -  If a "key_share" extension was supplied in the HelloRetryRequest,
      replacing the list of shares with a list containing a single
      KeyShareEntry from the indicated group.

Reply:
Ah, got it okay. That makes sense. I was thinking that extension was required for some reason. 

So would the change be to check to see if a key share extension was received in `s2n_server_hello_retry_recv`, and if it was, ensure that check (2) succeeds here, and if it wasn't received, only check to see if `conn->early_data_state == S2N_EARLY_DATA_REJECTED`?

Reply:
I think what I would do is gate line 71-107 with the check that the client actually received a keyshare extension. And then you can remove the early data rejected clause from the posix_ensure statement. 

It seems like we have a test for this situation: https://github.com/aws/s2n-tls/blob/main/tests/unit/s2n_client_early_data_indication_test.c#L412.
But I think this test is flawed, because it's not appropriately mocking how a server would send a HRR if it wanted to reject early data. The server in this case should write the HRR without the keyshare ext. 
Basically, additionally, I think we need to change the test so that the server excludes the keyshare ext.

Reply:
After looking at bouncy castle and openssl, it seems like implementations require a key share extension and don't provide an exception for early data. I therefor removed the exception, and fixed two failing tests that relied on not needing to update the selected_group: 36cf6a7

I also removed a test that checked for this specific exception that no longer exists: 75009a9